### PR TITLE
CHERKASY-61 RESTful webservice: Implement /studies/<studyId>/concepts/<c...

### PR DESCRIPTION
**Pull request by alchimystic**
...onceptId>/subjects

-implementation of PatientsResourceService.getPatients(OntologyTerm)
-integration tests

### CODE REVIEW

File: grails-app/services/org/transmartproject/db/querytool/PatientSetQueryBuilderService.groovy [hunk 1]
```diff
@@ -85,13 +82,24 @@ class PatientSetQueryBuilderService {
...
+                                ? " $databasePortabilityService.complementOperator "
+                                : ' INTERSECT ') +
                         "($panel.select)"
             }
         }
+    }
+
+    String buildPatientSetQuery(QtQueryResultInstance resultInstance,
+                                QueryDefinition definition)
+            throws InvalidArgumentsException {
```
Comment (cataphract) [inline grails-app/services/org/transmartproject/db/querytool/PatientSetQueryBuilderService.groovy]: Why did you change the throws clause?

File: src/groovy/org/transmartproject/db/ontology/AbstractQuerySpecifyingType.groovy [hunk 2]
```diff
@@ -56,4 +67,26 @@ abstract class AbstractQuerySpecifyingType implements MetadataSelectQuerySpecifi
...
+                                new Item(conceptKey: term.key)
+                        ]
+                )
+        ])
+
+        def patientsSql = patientSetQueryBuilderService.buildPatientIdListQuery(definition)
+        def patientsQuery = sessionFactory.currentSession.createSQLQuery patientsSql
+        def patientIdList = patientsQuery.list()
+
+        //this is a hack so integration tests work. for some reason the h2 schema doesn't have the right column type
```
Comment (cataphract) [inline src/groovy/org/transmartproject/db/ontology/AbstractQuerySpecifyingType.groovy]: Then fix the domain class.

File: transmart-core-db-tests/src/groovy/org/transmartproject/db/dataquery/TestDataHelper.groovy [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+        if (f != null) {
+            f.setAccessible(true) //useful later on
+            return f //found it
+        } else if (clazz != Object) {
+            return getField(clazz.superclass, name) //recurse into superclass, until we get the top
+        } else {
+            return null //no field
+        }
+    }
```
Comment (cataphract) [inline transmart-core-db-tests/src/groovy/org/transmartproject/db/dataquery/TestDataHelper.groovy:38]: Very java-ish, it would've have been easier to work with `groovy.lang.MetaBeanProperty`, but it's OK.

File: transmart-core-db-tests/src/groovy/org/transmartproject/db/dataquery/TestDataHelper.groovy [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            case Date:
+                return new Date()
+            default:
+                throw new UnsupportedOperationException("Not supported: $type.name. Care to add it?")
+        }
+    }
+
+    private static List<Field> getMandatoryFields(Class clazz) {
+        def list = clazz.constraints?.findAll({ it.value.nullable == false }) //get all not nullable properties
+        list.collect( {it.key} ).collect( { getField(clazz, it)}) //get the Field for them
```
Comment (cataphract) [inline transmart-core-db-tests/src/groovy/org/transmartproject/db/dataquery/TestDataHelper.groovy]: Leave a space after `{` and before `}` when defining `Closure`s (as you did on line 38 and on the second one on this line).

File: transmart-core-db-tests/src/groovy/org/transmartproject/db/dataquery/TestDataHelper.groovy [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ */
+class TestDataHelper {
+
+    /**
+     * Fills the object with dummy values for all the fields that are mandatory (nullable = false) and have no value
+     * @param clazz
+     * @param obj
+     */
+    static <T> void completeObject(Class<T> clazz, T obj) {
+        List<Field> fields = getMandatoryFields(clazz).findAll( {it.get(obj) == null} ) //all without value
```
Comment (cataphract) [inline transmart-core-db-tests/src/groovy/org/transmartproject/db/dataquery/TestDataHelper.groovy]: See comment on line 39.

File: transmart-core-db-tests/src/groovy/org/transmartproject/db/ontology/ConceptTestData.groovy [hunk 6]
```diff
@@ -39,4 +42,45 @@ class ConceptTestData {
...
+            o
+        }
+    }
+
+    /**
+     * @param list
+     * @return ConceptDimension list for the given I2b2 list
+     */
+    static List<ConceptDimension> createConceptDimensions(List<I2b2> list) {
+        list.collect( { new ConceptDimension(conceptPath: it.fullName, conceptCode: it.code)} )
```
Comment (cataphract) [inline transmart-core-db-tests/src/groovy/org/transmartproject/db/ontology/ConceptTestData.groovy]: three ws problems here

File: src/groovy/org/transmartproject/db/ontology/AbstractQuerySpecifyingType.groovy (hunk 2 above)
Comment (alchimystic) [inline src/groovy/org/transmartproject/db/ontology/AbstractQuerySpecifyingType.groovy]: I investigated this, and it seems to be a hibernate issue on BIGINT columns that are also identities. see 
http://stackoverflow.com/questions/18758347/hibernate-returns-bigintegers-instead-of-longs
http://jadimeo.wordpress.com/2009/09/05/sql-bigint-identity-columns-with-hibernate-annotations/
There are 2 proposed solutions:
-use Integer instead of Long
-implement our own IdentityGenerator and use it in this column.
I don't like either, and until we have a decision, i will leave this hack with some extra comments

File: transmart-core-db-tests/src/groovy/org/transmartproject/db/dataquery/TestDataHelper.groovy (hunk 3 above)
Comment (alchimystic) [inline transmart-core-db-tests/src/groovy/org/transmartproject/db/dataquery/TestDataHelper.groovy:38]: Will improve to use the MetaBeanProperty.
Indeed less code, specially avoiding the superclass field recursion.

**Comment by alchimystic** [comment]
All commented issues were fixed, except the Long to BigInteger mapping issue.
We will revisit this later, if needed
